### PR TITLE
varlen attention tutorial

### PR DESCRIPTION
**Pull request by liangel-02**
This PR is a tutorial for a variable length attention API

### CODE REVIEW

**Comment by drisspg** [comment]
@svekars any tips for reviewing this code in a more human friendly way?

**Comment by svekars** [comment]
There is more info here: https://github.com/pytorch/tutorials#contributing and there is a [script](https://gist.github.com/chsasank/7218ca16f8d022e02a9c0deb94a310fe) that will help you convert your .ipynb to .py

File: intermediate_source/variable_length_attention_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
+"""
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py:1]: please add to the toctree in the [compilers_index.rst](https://github.com/pytorch/tutorials/blob/main/compilers_index.rst?plain=1) and add a `.. customcarditem::`

File: intermediate_source/variable_length_attention_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+#    ``scaled_dot_product_attention``
+# -  Explore an example of how to use ``varlen_attn`` in a simple
+#    Transformer attention layer
+#
+# Prerequisites
+# ~~~~~~~~~~~~~
+#
+# -  PyTorch v2.10.0.dev or later
+# -  A basic understanding of attention and our current offerings. Please
+#    reference these tutorials for more details on flex attention and
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]: I think it probably needs to be capitalized: FlexAttention?
```suggestion
#    reference these tutorials for more details on FlexAttention and
```

File: intermediate_source/variable_length_attention_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+######################################################################
+#    | **Note:**
+#    | This tutorial currently requires you to use the PyTorch nightly
+#      build.
+#
+# What you will learn
+# ~~~~~~~~~~~~~~~~~~~
+#
+# -  Variable length attention and how it differs from
+#    ``scaled_dot_product_attention``
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]: Should we capitalize and add an abbreviation for clarity?
```suggestion
#    Scaled Dot-Product Attention (SDPA)
```

File: intermediate_source/variable_length_attention_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+# -------------------------------------
+#
+# In normal SDPA, sequences are expected to be a fixed length. In
+# practice, this means that input tensors are often **padded** to the same
+# length in a batch. However, this wastes both memory and compute through
+# storing this padding and performing unnecessary computations.
+#
+# Variable length attention handles sequences of varying length by
+# **packing** the tensors in a batch together and essentially collapsing
+# the batch dimension. Note that NestedTensor is another way to enable
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]:
```suggestion
# the batch dimension. Note that ``NestedTensor`` is another way to enable
```

File: intermediate_source/variable_length_attention_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+# variable length with packed tensors (see tutorial
+# `here <https://docs.pytorch.org/tutorials/unstable/nestedtensor.html>`__).
+#
+# However, we still need to maintain the boundaries between documents. To
+# do so, we compute cumulative sequence positions for query and key/value
+# that mark the end of documents. In the diagram below, doc 1 is 7 tokens
+# long, doc 2 is 10 tokens long, etc. so
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]: Something broken here?

```suggestion
# long, doc 2 is 10 tokens long, and so on. So 
```

File: intermediate_source/variable_length_attention_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+    main()
+
+
+######################################################################
+# Limitations
+# -----------
+#
+# Note that this op currently only works with NVIDIA CUDA on A100 machines
+# or newer. Supported dtypes include BF16 and FP16.
+#
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]: Need to add a Conclusion, something like:

```
In this tutorial, you have learned how to use the varlen_attn API in PyTorch to efficiently
handle sequences of varying lengths without padding. You explored how to create the
necessary metadata using cumulative sequence indices, implemented a simple
Transformer attention layer with variable length attention, and ran a complete
training step. This approach eliminates wasted computation on padding tokens
and enables more efficient training and inference for models processing multiple
documents of different lengths.

.. seealso::

* [nn.attention.varlen API](https://docs.pytorch.org/docs/stable/nn.attention.varlen.html)
* [Implementing High-Performance Transformers with Scaled Dot Product Attention ](https://docs.pytorch.org/tutorials/intermediate/scaled_dot_product_attention_tutorial.html)
* [torch.nn.functional.scaled_dot_product_attention](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html)

File: intermediate_source/variable_length_attention_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
+"""
+Using Variable Length Attention in PyTorch
+==========================================
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py:3]:
```suggestion
==========================================

.. meta::
   :description: Learn how to use PyTorch's varlen_attn API for efficient variable length attention without padding. Complete tutorial with code examples for training Transformers with packed sequences.
   :keywords: pytorch, attention, variable length, transformer, varlen_attn, packed sequences, memory optimization, torch.compile

**Author:** `Your Name <https://github.com/liangel-02>`_
```

File: intermediate_source/variable_length_attention_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+
+Summary
+-------
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]: Would just get rid of this.
```suggestion
```

File: intermediate_source/variable_length_attention_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+# Example
+# -------
+#
+# Let’s walk through a simple example of how we would use ``varlen_attn``
+# in the context of training a Transformer model.
+#
+
+
+######################################################################
+# Creating the metadata
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]: More descriptive titles will help with SEO.
```suggestion
# Creating Variable Length Metadata from Input Batches
```

File: intermediate_source/variable_length_attention_tutorial.py [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+    max_seqlen = 0
+    if len(all_seq_lengths) > 0:
+        all_seq_lengths = torch.cat(all_seq_lengths)
+        max_seqlen = all_seq_lengths.max().item()
+
+    return packed_cu_seqlens, max_seqlen
+
+
+######################################################################
+# Defining the Attention Block
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]:
```suggestion
#  Implementing Attention with ``varlen_attn``
```

File: intermediate_source/variable_length_attention_tutorial.py [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+#
+# -  Variable length attention and how it differs from
+#    ``scaled_dot_product_attention``
+# -  Explore an example of how to use ``varlen_attn`` in a simple
+#    Transformer attention layer
+#
+# Prerequisites
+# ~~~~~~~~~~~~~
+#
+# -  PyTorch v2.10.0.dev or later
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py:35]: Do we need to add hardware prerequisites like: NVIDIA A100 GPU or newer

File: intermediate_source/variable_length_attention_tutorial.py [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+# -  A basic understanding of attention and our current offerings. Please
+#    reference these tutorials for more details on flex attention and
+#    SDPA.
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]: Add a link to the SDPA tutorial https://docs.pytorch.org/tutorials/intermediate/scaled_dot_product_attention_tutorial.html

File: intermediate_source/variable_length_attention_tutorial.py [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+if __name__ == "__main__":
+    main()
+
+
+######################################################################
+# Limitations
+# -----------
+#
+# Note that this op currently only works with NVIDIA CUDA on A100 machines
+# or newer. Supported dtypes include BF16 and FP16.
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]: We can delete this and have it just in the prerequisites?

**Comment by svekars** [comment]
There is also an error here: https://github.com/pytorch/tutorials/actions/runs/19864783846/job/56924391840?pr=3660#step:9:4582 - let me check how should we test against 2.10.

File: intermediate_source/variable_length_attention_tutorial.py (hunk 5 above)
Comment (liangel-02) [reply intermediate_source/variable_length_attention_tutorial.py]: i think the rest of the sentence is on the next line

File: intermediate_source/variable_length_attention_tutorial.py (hunk 13 above)
Comment (liangel-02) [reply intermediate_source/variable_length_attention_tutorial.py]: cc @drisspg

**Comment by svekars** [comment]
@AlannaBurke can you please review this 2.10 tutorial.

**Comment by svekars** [comment]
Nightly preview: https://docs-preview.pytorch.org/pytorch/tutorials-nightly-preview/3660/intermediate/variable_length_attention_tutorial.html

File: intermediate_source/variable_length_attention_tutorial.py [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+# to handle sequences of varying lengths without padding. We explored how to create the
+# necessary metadata including the cumulative sequence indices, implemented a simple
+# Transformer attention layer with variable length attention, and ran a complete
+# training step.
+#
+# This approach eliminates wasted computation on padding tokens
+# and enables more efficient training and inference for models processing
+# documents of different lengths.
+
+"""
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]: Need to remove the """ as it's broken in the preview: https://docs-preview.pytorch.org/pytorch/tutorials-nightly-preview/3660/intermediate/variable_length_attention_tutorial.html#conclusion

And Out should not be there...

File: intermediate_source/variable_length_attention_tutorial.py [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+
+######################################################################
+# Overview of Variable Length Attention
+# -------------------------------------
+#
+# In normal SDPA, sequences are expected to be a fixed length. In
+# practice, this means that input tensors are often **padded** to the same
+# length in a batch. However, this wastes both memory and compute through
+# storing this padding and performing unnecessary computations.
+
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]: This part below is broken in the HTML preview: https://docs-preview.pytorch.org/pytorch/tutorials-nightly-preview/3660/intermediate/variable_length_attention_tutorial.html 

need to add a line of  #### after each empty line

File: intermediate_source/variable_length_attention_tutorial.py [hunk 16]
```diff
@@ -0,0 +1,356 @@
...
+#        value: torch.Tensor,
+#        cu_seq_q: torch.Tensor,
+#        cu_seq_k: torch.Tensor,
+#        max_q: int,
+#        max_k: int,
+#        is_causal: bool = False,
+#        return_aux: AuxRequest | None = None,
+#    ) -> torch.Tensor | tuple[torch.Tensor, torch.Tensor]:
+
+# ``query``, ``key``, and ``value`` correspond to the ``q``, ``k``, and
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py:106]: Same as above, need to ad a # in the beginning of the empty line: https://docs-preview.pytorch.org/pytorch/tutorials-nightly-preview/3660/intermediate/variable_length_attention_tutorial.html#definition

File: intermediate_source/variable_length_attention_tutorial.py [hunk 17]
```diff
@@ -0,0 +1,356 @@
...
+#        key: torch.Tensor,
+#        value: torch.Tensor,
+#        cu_seq_q: torch.Tensor,
+#        cu_seq_k: torch.Tensor,
+#        max_q: int,
+#        max_k: int,
+#        is_causal: bool = False,
+#        return_aux: AuxRequest | None = None,
+#    ) -> torch.Tensor | tuple[torch.Tensor, torch.Tensor]:
+
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py]:
```suggestion
#
```

File: intermediate_source/variable_length_attention_tutorial.py [hunk 18]
```diff
@@ -0,0 +1,356 @@
...
+# ``query``, ``key``, and ``value`` correspond to the ``q``, ``k``, and
+# ``v`` of the packed input. ``cu_seq_q`` and ``cu_seq_k`` are the
+# cumulative indices for query and key/value, respectively. These mark the
+# logical boundaries that separate the documents in our input. ``max_q``
+# and ``max_k`` are the maximum sequence lengths of query and key,
+# respectively. ``is_causal`` applies causal masking if set to True and
+# ``return_aux`` specifies which auxiliary outputs to return (ie ``lse``).
+
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py:113]:
```suggestion
#
```

File: intermediate_source/variable_length_attention_tutorial.py [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+# In normal SDPA, sequences are expected to be a fixed length. In
+# practice, this means that input tensors are often **padded** to the same
+# length in a batch. However, this wastes both memory and compute through
+# storing this padding and performing unnecessary computations.
+
+######################################################################
+# Variable length attention handles sequences of varying length by
+# **packing** the tensors in a batch together and essentially collapsing
+# the batch dimension.
+
```
Comment (svekars) [inline intermediate_source/variable_length_attention_tutorial.py:54]: yes the "Variable length attention" part got fixed but the para below still shows as code block: https://docs-preview.pytorch.org/pytorch/tutorials-nightly-preview/3660/intermediate/variable_length_attention_tutorial.html#overview-of-variable-length-attention

need to make sure, every empty line starts with a #. if  you add an empty line the new line needs to have ###### and then your text. The figure below is not showing too. Here is more about sphinx-gallery syntax: https://sphinx-gallery.github.io/stable/syntax.html#embed-rest-in-your-example-python-files

File: _static/img/varlen_diagram.png
Comment (AlannaBurke) [inline _static/img/varlen_diagram.png:1]: Is there any way to make this diagram accessible? If not, can you make sure that there's a thorough description of it?

Comment (liangel-02) [reply _static/img/varlen_diagram.png:1]: added more alt text!

Comment (AlannaBurke) [reply _static/img/varlen_diagram.png:1]: excellent, thanks!

**Comment by AlannaBurke** [comment]
looks great, make sure you've resolved all of @svekars 's comments

**Comment by liangel-02** [comment]
> looks great, make sure you've resolved all of @svekars 's comments

yep resolved
